src/sinks/humio_logs.rs

```diff
@@ -118,3 +118,156 @@ mod tests {
+
+    #[test]
+    fn humio_insert_message() {
+        let mut rt = runtime();
+        let cx = SinkContext::new_test();
+
+        rt.block_on_std(async move {
+            let repo = create_repository().await;
+
+            let config = config(&repo.default_ingest_token);
+
+            let (sink, _) = config.build(cx).unwrap();
+
+            let message = random_string(100);
+            let event = Event::from(message.clone());
+
+            sink.send(event).compat().await.unwrap();
+
+            let entry = find_entry(repo.name.as_str(), message.as_str()).await;
+
+            assert_eq!(message, entry["message"].as_str().unwrap());
+            assert!(
+                entry.get("@error").is_none(),
+                "Humio encountered an error parsing this message: {}",
+                entry
+                    .get("@error_msg")
+                    .unwrap_or(&json!("no error message"))
+            );
+        });
+    }
+
+    #[derive(Debug)]
+    struct HumioRepository {
+        name: String,
+        default_ingest_token: String,
+    }
+
+    /// create a new test config with the given ingest token
+    fn config(token: &str) -> super::HumioLogsConfig {
+        super::HumioLogsConfig {
+            host: Some(HOST.to_string()),
+            token: token.to_string(),
+            compression: Compression::None,
+            encoding: Encoding::Json.into(),
+            batch: BatchConfig {
+                max_events: Some(1),
+                ..Default::default()
+            },
+            ..Default::default()
+        }
+    }
+
+    /// create a new test humio repository to publish to
+    async fn create_repository() -> HumioRepository {
+        let client = reqwest::Client::builder().build().unwrap();
+
+        // https://docs.humio.com/api/graphql/
+        let graphql_url = format!("{}/graphql", HOST);
+
+        let name = random_string(50);
+
+        let params = json!({
+        "query": format!(
+            r#"
+mutation {{
+  createRepository(name:"{}") {{
+    repository {{
+      name
+      type
+      ingestTokens {{
+        name
+        token
+      }}
+    }}
+  }}
+}}
+"#,
+            name
+        ),
+        });
+
+        let res = client
+            .post(&graphql_url)
+            .json(&params)
+            .send()
+            .await
+            .unwrap();
+
+        let json: JsonValue = res.json().await.unwrap();
+        let repository = &json["data"]["createRepository"]["repository"];
+
+        let token = repository["ingestTokens"].as_array().unwrap()[0]["token"]
+            .as_str()
+            .unwrap()
+            .to_string();
+
+        HumioRepository {
+            name: repository["name"].as_str().unwrap().to_string(),
+            default_ingest_token: token,
+        }
+    }
+
+    /// fetch event from the repository that has a matching message value
+    async fn find_entry(repository_name: &str, message: &str) -> serde_json::value::Value {
```

Comment:
Does the returned value have a fixed structure? Since we are testing the contents of the structure above, that could make it worth making a concrete type for it. Of course, if it doesn't have a fixed structure, that won't work.

Reply:
:+1: it is semi-structured. I'll add a type.

Reply:
Added in https://github.com/timberio/vector/pull/3260/commits/c2e6a3a0291baeec1f321cf06eb17783415403fb